Dev

carlobortolan opened the pull request:
closes #8 user authentication (PR #14)
closes #31 initial UI views (PR #53)
closes #32 implement custom errors (PR #53)
closes #54 set up grpc client (PR #59)

carlobortolan committed:
8 user authentication (#14)

* Update README.md and dependencies

* Add basic app structure and assets

* Implement first models, viewmodels and api calls needed for user auth

* Implement TUM-SSO authentication and update README.md and

* Clean up auth_handler.dart and update routes

* Update userState and userViewModel to use riverpod and rxdart

* Refactor code to improve readability

* Remove unused import

* Clean up code, refactor models, update error-handling

Anishyou committed:
31 initial UI views (#53)

* Basic tum login functionality, not yet done internal account

* pushin android and yaml files also

* Added some more views,course overview and setting

* Added mycourses and public courses views,also solved navigation bug

* added bottom bar functionality with solving navigation using route=false

* removed views folder

* changed names of notification file

* Solved navigation bug from notification screen

* Added some photos,cleared login screen and add same on 1 screen and fixed bottom nav bar

* Added reusable widgets,constants,corrrected name for bookmark

* changed bookmarks text to pinned

* added login functionality

* solved login bug

---------

Co-authored-by: ge59dil <anish.goyal@tum.de>

carlobortolan committed:
32 implement custom errors (#55)

* Rename View->views and ViewModel -> view_model

* Run `dart fix --apply && dart format ./lib`

* Implement error handling for http requests&responses

carlobortolan committed:
54 set up grpc client (#59)

* Rename `lib/model` -> `lib/models` and `networking/apis` -> `networking/api`

* Update error handling to work with gRPC codes

* Setup proto dir structure

* Implement grpc_handler.dart

* Add (de-)serialization from/to proto

* Implement `fetchUser()` gRPC method in `user_handler.dart` and update `user_viewmodel.dart`

* Add doc

* Fix args & linting in model_generator.dart

* Update dependencies

* Update README.md

* Update `routes.dart` to automatically configure root url

* Update doc

* Add `setup_grpc.sh`

carlobortolan committed:
Tmp (#69)

Update workflow with right bundle id

Co-authored-by: Jakob Körber <56073945+jakobkoerber@users.noreply.github.com>

jakobkoerber committed:
Merge branch 'main' into dev

jakobkoerber commented on lib/views/enablenotification_screen.dart:
```diff
@@ -0,0 +1,73 @@
+              const SizedBox(height: 48),
+              const Spacer(),
+              ElevatedButton(
+                style: ElevatedButton.styleFrom(
```
Add this to a theme 😄

GravityDarkLab replied:
done ✅  see this [commit](https://github.com/TUM-Dev/gocast_mobile/commit/e32ce3fa01ea8b05f6013cd2b31d9bce6da0abdd#diff-7eb87a0c9553264ec1d72f99b88c47c68e4be65ff1ddeccf4d09b8da09a28980)

p.s: Theme will be improved

jakobkoerber commented on lib/views/internallogin_screen.dart:
```diff
@@ -0,0 +1,134 @@
+class InternalloginScreen extends ConsumerWidget {
+  const InternalloginScreen({super.key});
+
+  Future<void> handleBasicLogin(
```
Seems to be a case for ViewModel

GravityDarkLab replied:
This will be done in this [branche](https://github.com/TUM-Dev/gocast_mobile/tree/34-clean-code)

jakobkoerber commented on lib/views/internallogin_screen.dart:
```diff
@@ -0,0 +1,134 @@
+    final TextEditingController usernameController = TextEditingController();
+    final TextEditingController passwordController = TextEditingController();
+
+    return Scaffold(
```
Extract some code into function or widgets, the function body is quite long

GravityDarkLab replied:
This will be done in this [branche](https://github.com/TUM-Dev/gocast_mobile/tree/34-clean-code)

jakobkoerber reviewed (commented):
Hey @TUM-Dev/gocast-ipraktikum, great work so far! As requested, I added some comments and suggestions to your code! Feel free to consider and implement them. Happy coding!

Anishyou commented on lib/views/settings_screen.dart:
```diff
@@ -0,0 +1,142 @@
+            ),
+            onTap: () {
+              setState(() {
+                isDarkMode = !isDarkMode;
```
yes it is to be implemented later.

Anishyou commented on lib/views/settings_screen.dart:
```diff
@@ -0,0 +1,142 @@
+            ),
+            onTap: () {
+              setState(() {
+                isPushNotificationsEnabled = !isPushNotificationsEnabled;
```
yes, to do this later

carlobortolan commented on lib/bootstrap.dart:
```diff
@@ -0,0 +1,33 @@
+import 'package:bloc/bloc.dart';
+import 'package:flutter/widgets.dart';
+
+class AppBlocObserver extends BlocObserver {
```
No, there's no reason behind that. We actually only plan to use `riverpod`. `boostrap.dart` seems to be dead code left behind by our initial GitLab project setup.
Will be removed in a future PR.

jakobkoerber replied:
I would recommend removing it now or creating an issue to prevent it from being forgotten and ending up as dead code 😄

GravityDarkLab replied:
done ✅

carlobortolan commented on lib/view_models/user_viewmodel.dart:
```diff
@@ -0,0 +1,48 @@
+
+  Future<void> _fetchUser() async {
+    await UserHandler(_grpcHandler).fetchUser().then(
+          (value) => current.value.setUser(value),
```
Yes. (At least it currently does so after the client successfully received a JWT, calls the `getUser` endpoint to set the current user and then switches to the start view)

Also see `user_state_model.dart`:
```dart
  void setUser(User newUser) {
    debugPrint("User set to: ${newUser.name}");
    user = newUser;
  }
```
Console log when logging in: (obviously for debug purposes only)
![image](https://github.com/TUM-Dev/gocast_mobile/assets/106114526/fe42f5f4-7441-4ab0-a9ba-67bbd5a3c784)

carlobortolan commented on lib/models/course/bookmark_model.dart:
```diff
@@ -0,0 +1,35 @@
+import 'package:gocast_mobile/base/networking/api/gocast/api_v2.pb.dart'
+    as proto;
+
+class Bookmark {
```
That's a good point :sweat_smile: ... the only reason for this that I can think of was that we already implemented some models before implementing / deciding to use gRPC and therefore just continued using the old models.
We'll make sure to replace our old models with the proto models this in a future PR.

GravityDarkLab replied:
Creating a model from a proto provides:
- Abstraction
- Decouples the app logic from the proto implementation
- Null Safety and Error Handling

jakobkoerber replied:
Hmm, I get what you mean, but this not really accurate in this case. gRPC already provides null and type safety. If one needs to extends the generated models, an extension can be used. This basically creates a duplicate which can ultimately lead to confusion for other developers.

carlobortolan replied:
>  We'll make sure to replace our old models with the proto models in a future PR.

Implemented in [this commit](https://github.com/TUM-Dev/gocast_mobile/commit/188335bbcc0f1ccf65cbca3e59508862fb7f045d).

jakobkoerber replied:
Looks good! I would suggest, if it's done, to merge it into dev and include it in this PR

carlobortolan commented on lib/models/course/course_model.dart:
```diff
@@ -0,0 +1,26 @@
+  int year;
+
+  // Factory method to create a Course instance from the gRPC response
+  factory Course.fromProto(proto.Course course) {
```
(see comment above)

carlobortolan commented on lib/models/user/user_model.dart:
```diff
@@ -0,0 +1,63 @@
+  List<Bookmark> bookmarks;
+
+  // Factory method to create a User instance from the gRPC response
+  factory User.fromProto(proto.User user) {
```
(see comment above)

carlobortolan commented on lib/models/token_model.dart:
```diff
@@ -0,0 +1,43 @@
+  static Future<void> saveToken(String key, List<Cookie> cookies) async {
+    for (var cookie in cookies) {
+      if (cookie.name == key) {
+        SharedPreferences prefs = await SharedPreferences.getInstance();
```
Thanks for the info, will implement this in a [future PR](https://github.com/TUM-Dev/gocast_mobile/issues/70) :+1:.

GravityDarkLab replied:
done ✅

GravityDarkLab replied:
done ✅

GravityDarkLab commented on devtools_options.yaml:
```diff
@@ -0,0 +1 @@
+extensions:
```
this is a file where you can customise your developer tools 

- Theme Customization
- Network Profiling:
- Layout Customization:
......

But can be deleted if you say so :)

jakobkoerber replied:
You definitely don't have to delete it 😄 I was just interested because I haven't heard of it before

GravityDarkLab commented on lib/views/utils/course_card_view.dart:
```diff
@@ -0,0 +1,48 @@
+  @override
+  Widget build(BuildContext context) {
+    return Container(
+      width: 160, // Adjust the width as needed
```
This will be done in this [branche](https://github.com/TUM-Dev/gocast_mobile/tree/34-clean-code)

GravityDarkLab commented on lib/views/utils/course_screen.dart:
```diff
@@ -0,0 +1,68 @@
+          ],
+        ),
+      ),
+      bottomNavigationBar: const CustomBottomNavBar(),
```
This will be done in this [branche](https://github.com/TUM-Dev/gocast_mobile/tree/34-clean-code)

GravityDarkLab commented on lib/views/utils/courselist_screen.dart:
```diff
@@ -0,0 +1,42 @@
+      body: ListView(
+        children: videoCards,
+      ),
+      bottomNavigationBar: const CustomBottomNavBar(),
```
This will be done in this [branche](https://github.com/TUM-Dev/gocast_mobile/tree/34-clean-code)

GravityDarkLab commented on lib/views/utils/courselist_screen.dart:
```diff
@@ -0,0 +1,42 @@
+
+class CourseListScreen extends ConsumerWidget {
+  final String title;
+  final List<Widget> videoCards;
```
This will be done in this [branche](https://github.com/TUM-Dev/gocast_mobile/tree/34-clean-code)

GravityDarkLab commented on lib/views/welcome_screen.dart:
```diff
@@ -0,0 +1,150 @@
+
+class WelcomeScreen extends ConsumerWidget {
+  const WelcomeScreen({super.key});
+  Future<void> handleSSOLogin(
```
This will be done in this [branche](https://github.com/TUM-Dev/gocast_mobile/tree/34-clean-code)

GravityDarkLab commented on lib/views/welcome_screen.dart:
```diff
@@ -0,0 +1,150 @@
+
+  @override
+  Widget build(BuildContext context, WidgetRef ref) {
+    final TextEditingController usernameController = TextEditingController();
```
This will be done in this [branche](https://github.com/TUM-Dev/gocast_mobile/tree/34-clean-code)

GravityDarkLab commented on lib/views/welcome_screen.dart:
```diff
@@ -0,0 +1,150 @@
+    TextEditingController passwordController,
+  ) {
+    return ElevatedButton(
+      style: ElevatedButton.styleFrom(
```
done ✅  see this [commit](https://github.com/TUM-Dev/gocast_mobile/commit/e32ce3fa01ea8b05f6013cd2b31d9bce6da0abdd#diff-7eb87a0c9553264ec1d72f99b88c47c68e4be65ff1ddeccf4d09b8da09a28980)

GravityDarkLab commented on lib/views/settings_screen.dart:
```diff
@@ -0,0 +1,142 @@
+          ),
+          ListTile(
+            title: const Text('Push notifications'),
+            trailing: CupertinoSwitch(
```
This will be done in this [branche](https://github.com/TUM-Dev/gocast_mobile/tree/34-clean-code)

GravityDarkLab commented on lib/views/utils/course_screen.dart:
```diff
@@ -0,0 +1,68 @@
+
+class CoursesScreen extends StatelessWidget {
+  final String title;
+  final List<Widget> courseCards;
```
This will be done in this [branche](https://github.com/TUM-Dev/gocast_mobile/tree/34-clean-code)

GravityDarkLab commented on lib/views/notifications_screen.dart:
```diff
@@ -0,0 +1,58 @@
+          );
+        },
+      ),
+      bottomNavigationBar: const CustomBottomNavBar(),
```
This will be done in this [branche](https://github.com/TUM-Dev/gocast_mobile/tree/34-clean-code)

GravityDarkLab commented on lib/views/courseoverview_screen.dart:
```diff
@@ -0,0 +1,117 @@
+          ],
+        ),
+      ),
+      bottomNavigationBar: const CustomBottomNavBar(),
```
This will be done in this [branche](https://github.com/TUM-Dev/gocast_mobile/tree/34-clean-code)

jakobkoerber commented:
I would suggest, if it's done, to merge the [other branch](https://github.com/TUM-Dev/gocast_mobile/tree/34-clean-code) into dev and include it in this PR, therefore only reviewed and improved code is merged onto main 😄

jakobkoerber committed:
Update beta release workflow

GravityDarkLab committed:
34 clean code base (#83)

* Add config directory for better structure

* Refactor ModelGenerator

* Add Logging in handlers

* Add Logging in handlers

* Add secure storage for the token

* removed Bootstrap

* Add secure storage for the token

* Update `main.dart` to use `StreamBuilder` instead of `StreamProvider`

* Refactor  UserViewModel

* Replace old models with proto models

* Fix linting

* Improve code structure

* Add Theme

* Refactor 'internal_login_screen.dart'

* Refactor

* Refactor: Remove unused TextEditingController instances in WelcomeScreen

- Eliminated declarations of usernameController and passwordController in WelcomeScreen as they were not used.
- Ensured clean and optimized code by removing unnecessary Riverpod provider references.
- Improved overall code readability and maintainability.

* Optimize CourseOverview to Prevent Unnecessary Re-renders

- Resolved issue causing re-renders of AppBar and BottomNavigationBar on tab changes.
- Implemented Column layout to display 'My Courses' and 'Public Courses' sections together, ensuring consistent visibility.
- Added spacing between course sections for better visual clarity and user experience.
- Enhanced layout performance and overall responsiveness of the CourseOverview screen.

* Add code documentation for course_overview and course_section

* Refactor Course Sections for Reusability and Clarity

- Introduced CourseContentScreen to centralize the display logic of course sections.
- Refactored PinnedCourses and DownloadsScreen to use CourseContentScreen, enhancing code reusability.
- Added in-code documentation for clarity and future maintainability.

* Fix INT64 type

* Fix Linting

* Fix colors on TabBar not changing correctly

* Added list view builder in course_view

* changed list view to list view builder in course_view and addressed some trailing commas issue

* added material design switch in setting

* deleteing course outline

* Added base_view for app bar and bottom navigatoe

* Added base_view in course_Screen

* added onViewAll functionality

* no re rendering when click on same page

* Improve the usage of the user Model view.
- Added loading state when clicked on login
    - `handleSSOLogin()`
    - `handleSSOLogin()`
    - Added `BehaviorSubject<bool> isLoading = BehaviorSubject.seeded(false);`

* Refactor `course_card_view.dart` and added documentation for `base_view.dart`, `course_card_view.dart`, `course_overview_view.dart`

* Refactor `ssoAuth` and added documentation

* Refactor `SettingsScreen` and added documentation to `internal_login_view` and `welcome_screen_view`

* Fix in `ssoAuth` when ` Navigator.pushNamed(context, '/home');` called to set the loading state of login to false.

* Fix in `ssoAuth` when ` Navigator.pushNamed(context, '/home');` called to set the loading state of login to false.

* Rename `routes.dart` to `app_config.dart` and moved to the `config` package

* **Refactored** `UserViewModel` and Login Handling**: Moved login logic and `TextEditingController`s into `UserViewModel`

* ## Refactor: Integrate StateNotifier in UserViewModel
- Refactored `UserViewModel` to extend `StateNotifier<UserState>`.
- Eliminated separate `isLoading` BehaviorSubject in `UserViewModel`. All state changes, including loading states, are now handled directly through the unified `UserState` managed by the notifier.

* ## Refactor: Courses views to have `List<VideoCard>` or `List<VideoCard>

* ## Refactor: Courses views to have to make better use of `List.builder` and added a temporary `CourseModel` in course_model.dart that will be replaced once API courses retrieving works.

---------

Co-authored-by: carlobortolan <carlobortolan@gmail.com>
Co-authored-by: ge59dil <anish.goyal@tum.de>

GravityDarkLab reviewed (dismissed):
Issue: #34  implemented most of Jakob's feedbacks and he already approved to merge it into dev.

jakobkoerber commented on lib/views/on_boarding_view/welcome_screen_view.dart:
```diff
@@ -0,0 +1,136 @@
+      ),
+      child: const Text('Continue without', style: TextStyle(fontSize: 18)),
+      onPressed: () {
+        //TODO: Continue without login action
```
Last suggestion: is it possible to implement this right now? Otherwise the TestFlight version is not really usable beyond the WelcomeScreen

carlobortolan replied:
Should be no problem - I'll have a look at it now

jakobkoerber replied:
Sounds good as a temporary solution. Once the API is deployed things get much clearer I suppose

carlobortolan committed:
Add redirect after "Continue Without" (#87)

GravityDarkLab reviewed (approved):
seems good to me
